### PR TITLE
:technologist: Min Heap --- `add`

### PR DESCRIPTION
**Pull request by jameshughes89**
### What
1. Write the `add` method

Some private methods too 

2. `expandCapacity`
3. `bubbleUp`
4. `swap`
5. `getParent`

### Testing
:+1:  but no test for ordering yet as, unless there is a assertJ trick to that, I would need `remove` for the remove ordering tests. 

### Additional Notes
Not sure how else the `return` would look on `add`

### CODE REVIEW

File: src/main/java/ArrayMinHeap.java [hunk 1]
```diff
@@ -26,7 +26,41 @@ public ArrayMinHeap(int initialCapacity) {
...
     public boolean add(T element) {
-        return false;
+        if (size() == heap.length) {
+            expandCapacity();
+        }
+        heap[size] = element;
+        bubbleUp(size);
+        size++;
+        return true;
+    }
```
Comment (jameshughes89) [inline src/main/java/ArrayMinHeap.java]: Idk what the `return false` case would be here.

Comment (twentylemon) [reply src/main/java/ArrayMinHeap.java]: There isn't always a `false` case. `add` returns true if the item was added, if the collection was changed. In this case, the collection _always_ changes, the element can always be added, so `return true`.

**Comment by twentylemon** [comment]
> 👍 but no test for ordering yet as

[containsExactly](https://assertj.github.io/doc/#assertj-core-group-contains) means those elements in that order, but it relies on the `iterator()` again.

File: src/main/java/ArrayMinHeap.java [hunk 2]
```diff
@@ -26,7 +26,41 @@ public ArrayMinHeap(int initialCapacity) {
...
+
+    @SuppressWarnings("unchecked")
+    private void expandCapacity() {
+        T[] newHeap = (T[]) new Comparable[heap.length * 2];
+        System.arraycopy(heap, 0, newHeap, 0, heap.length);
```
Comment (twentylemon) [inline src/main/java/ArrayMinHeap.java]: Don't use `arraycopy`. [Arrays#copyOf](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Arrays.html#copyOf(T%5B%5D,int)) exists for exactly this purpose.

**Comment by jameshughes89** [comment]
> > +1 but no test for ordering yet as
> 
> [containsExactly](https://assertj.github.io/doc/#assertj-core-group-contains) means those elements in that order, but it relies on the `iterator()` again.

I'll have you know I specifically noted that one.

File: src/test/java/ArrayMinHeapTest.java [hunk 3]
```diff
@@ -33,10 +34,51 @@ void isEmpty_empty_returnsTrue() {
         @Nested
         class WhenSingleton {
 
+            @BeforeEach
+            void addSingleInteger() {
+                classUnderTest.add(10);
+                preState.add(10);
+            }
```
Comment (twentylemon) [inline src/test/java/ArrayMinHeapTest.java:47]: You'll still want tests for `add`. When empty/singleton/whatever, adding something returns true. You'd probably also want to check equality, but other tests will be doing that anyways, so you at least want to check the return value.

**Review by twentylemon — APPROVED** [review]

File: src/main/java/ArrayMinHeap.java (hunk 2 above)
Comment (jameshughes89) [reply src/main/java/ArrayMinHeap.java]: I ended up
1. removing `expandCapacity` altogether for just the `Arrays#copyOf`
2. Added a test for `WhenLarge`

File: src/test/java/ArrayMinHeapTest.java [hunk 4]
```diff
@@ -25,19 +25,113 @@ void size_empty_returnsZero() {
...
+            assertThat(classUnderTest.add(10))
+                    .isTrue();
+        }
+
+        @Test
+        void add_empty_updatesHeap() {
+            classUnderTest.add(10);
+            assertThat(classUnderTest)
+                    .isNotEqualTo(preState);
+        }
```
Comment (jameshughes89) [inline src/test/java/ArrayMinHeapTest.java]: would it be _better_ to test that it is not equal to an new & empty queue under the nested `Singleton` test class?

File: src/test/java/ArrayMinHeapTest.java [hunk 5]
```diff
@@ -25,19 +25,113 @@ void size_empty_returnsZero() {
...
+                assertThat(classUnderTest.add(10))
+                        .isTrue();
+            }
+
+            @Test
+            void add_singleton_updatesHeap() {
+                classUnderTest.add(10);
+                assertThat(classUnderTest)
+                        .isNotEqualTo(preState);
+            }
```
Comment (jameshughes89) [inline src/test/java/ArrayMinHeapTest.java]: overkill?

File: src/test/java/ArrayMinHeapTest.java [hunk 6]
```diff
@@ -25,19 +25,113 @@ void size_empty_returnsZero() {
...
+                    assertThat(classUnderTest.add(10))
+                            .isTrue();
+                }
+
+                @Test
+                void add_many_updatesHeap() {
+                    classUnderTest.add(10);
+                    assertThat(classUnderTest)
+                            .isNotEqualTo(preState);
+                }
```
Comment (jameshughes89) [inline src/test/java/ArrayMinHeapTest.java]: Same as above --- overkill?

File: src/test/java/ArrayMinHeapTest.java (hunk 4 above)
Comment (twentylemon) [reply src/test/java/ArrayMinHeapTest.java]: I moreso meant asserting `thing.add(x) == true`. Asserts like this are also testing `equals`. More of a workflow test

File: src/test/java/ArrayMinHeapTest.java [hunk 7]
```diff
@@ -25,19 +25,113 @@ void size_empty_returnsZero() {
...
         void isEmpty_empty_returnsTrue() {
             assertThat(classUnderTest.isEmpty())
                     .isTrue();
         }
 
+        @Test
+        void add_empty_returnsTrue() {
+            assertThat(classUnderTest.add(10))
+                    .isTrue();
+        }
```
Comment (jameshughes89) [inline src/test/java/ArrayMinHeapTest.java:38]: [Is this what you meant?](https://github.com/jameshughes89/cs201/pull/13#discussion_r1021976600)

Comment (twentylemon) [reply src/test/java/ArrayMinHeapTest.java:38]: 1

**Comment by jameshughes89** [comment]
1. Removed the iterator from the class because (a) don't really care and (b) it was messing with `.usingRecursiveComparison()`.
2. ~By using `.usingRecursiveComparison()`,~ The equality check it's not so much about checking equality, but checking that the add actually changed something internally.

**Pull request merged**